src/System.Net.Primitives/src/System/Net/CookieContainer.cs

```diff
@@ -0,0 +1,1062 @@
+                    if (pathList == null)
+                    {
+                        pathList = new PathList();
+                        AddRemoveDomain(cookie.DomainKey, pathList);
+                    }
+                }
+                int domain_count = pathList.GetCookiesCount();
+
+                CookieCollection cookies;
+                lock (pathList.SyncRoot)
+                {
+                    cookies = (CookieCollection)pathList[cookie.Path];
+
+                    if (cookies == null)
+                    {
+                        cookies = new CookieCollection();
+                        pathList[cookie.Path] = cookies;
+                    }
+                }
+
+                if (cookie.Expired)
+                {
+                    // Explicit removal command (Max-Age == 0)
+                    lock (cookies)
+                    {
+                        int idx = cookies.IndexOf(cookie);
+                        if (idx != -1)
+                        {
+                            cookies.RemoveAt(idx);
+                            --_count;
+                        }
+                    }
+                }
+                else
+                {
+                    // This is about real cookie adding, check Capacity first
+                    if (domain_count >= _maxCookiesPerDomain && !AgeCookies(cookie.DomainKey))
+                    {
+                        return; // Cannot age: reject new cookie
+                    }
+                    else if (_count >= _maxCookies && !AgeCookies(null))
+                    {
+                        return; // Cannot age: reject new cookie
+                    }
+
+                    // About to change the collection
+                    lock (cookies)
+                    {
+                        _count += cookies.InternalAdd(cookie, true);
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                if (e is OutOfMemoryException)
+                {
+                    throw;
+                }
+
+                if (throwOnError)
+                {
+                    throw new CookieException(SR.net_container_add_cookie, e);
+                }
+            }
+        }
+
+        // This function, when called, must delete at least one cookie.
+        // If there are expired cookies in given scope they are cleaned up.
+        // If nothing is found the least used Collection will be found and removed
+        // from the container.
+        //
+        // Also note that expired cookies are also removed during request preparation
+        // (this.GetCookies method).
+        //
+        // Param. 'domain' == null means to age in the whole container.
+        private bool AgeCookies(string domain)
+        {
+            // Border case: shrunk to zero
+            if (_maxCookies == 0 || _maxCookiesPerDomain == 0)
+            {
+                _domainTable = new Hashtable();
```

Comment:
I'm not quite understanding the concurrency guarantees of these types.  In some places, we're locking on ```_domainTable.SyncRoot```, but here we're replacing the entire ```_domainTable``` object, without a lock, such that other threads could still be accessing the old _domainTable.  Is that ok?